### PR TITLE
drivers: timer: extend nrf_rtc_timer to handle overflow

### PR DESCRIPTION
**Pull request by pkwiek**
nrf_rtc_timer is extended with a capability to handle RTC overflow,
allowing it to operate on absolute RTC ticks, rather than relative
ticks. The overflow handling is ZLI-proof, due to use of exclusive
access instructions instead of Zephyr's synchronization primitives.
Callbacks are given an absolute RTC tick value as a parameter instead
of CC register's value.

Signed-off-by: Jedrzej Ciupis <jedrzej.ciupis@nordicsemi.no>
Signed-off-by: Paweł Kwiek <pawel.kwiek@nordicsemi.no>

### CODE REVIEW

File: tests/drivers/timer/nrf_rtc_timer/src/main.c [hunk 1]
```diff
@@ -53,34 +54,48 @@ static void stop_zli_timer0(void)
...
-	uint32_t now = z_nrf_rtc_timer_read();
-	uint32_t diff = (now - cc_value) & 0x00FFFFFF;
+	uint64_t now = z_nrf_rtc_timer_read();
+	uint64_t diff = (now - target_time);
 
 	zassert_true(diff <= data->delay,
-		"Handler called in wrong time (%d), set cc: %d, got cc: %d",
-		now, data->cc_val, cc_value);
+		"Handler called in wrong time (%d), set time: %d, got time: %d",
+		now, data->target_time, target_time);
```
Comment (ankuns) [inline tests/drivers/timer/nrf_rtc_timer/src/main.c:79]: `%d` format string does not match `uint64_t`

File: drivers/timer/nrf_rtc_timer.c [hunk 2]
```diff
@@ -238,12 +431,56 @@ static void sys_clock_timeout_handler(int32_t chan,
...
+		bool shall_execute_handler = false;
+		void *user_context;
+		uint32_t mcu_critical_state;
+
+		event_clear(chan);
+		event_disable(chan);
+		overflow_and_counter_get(&overflow, &counter);
+
+		curr_time =
+		    overflow_and_counter_to_target_time(overflow, counter);
```
Comment (ankuns) [inline drivers/timer/nrf_rtc_timer.c:457]: You could defer this calculation until `handler` is called.
line 480:
```c
if (shall_execute_handler) {
    uint64_t t = overflow_and_counter_to_target_time(overflow, counter);
    handler(chan, t, user_context);
}

File: drivers/timer/nrf_rtc_timer.c [hunk 3]
```diff
@@ -238,12 +431,56 @@ static void sys_clock_timeout_handler(int32_t chan,
...
+		/* If target_cc is in the past or is equal to current counter
+		 * value, execute the handler.
+		 */
+		if (counter_sub(cc_data[chan].target_cc, counter + 1)
+		    > COUNTER_HALF_SPAN) {
+			handler = cc_data[chan].callback;
+			user_context = cc_data[chan].user_context;
+			cc_data[chan].callback = NULL;
+
+			shall_execute_handler = handler != NULL;
```
Comment (ankuns) [inline drivers/timer/nrf_rtc_timer.c:475]: You can give up with separate variable `shall_execute_handler`.
Let's simply have initial value in line 447
```c
z_nrf_rtc_timer_compare_handler_t handler = NULL
```
then
```c
if (handler)
{
    // ... 
    handler()
}
```

File: drivers/timer/nrf_rtc_timer.c [hunk 4]
```diff
@@ -238,12 +431,56 @@ static void sys_clock_timeout_handler(int32_t chan,
...
+		 * cc_data structure and prevent higher priority contexts
+		 * (including ZLIs) from overwriting it.
+		 */
+		mcu_critical_state = __get_PRIMASK();
+		__disable_irq();
+
+		/* If target_cc is in the past or is equal to current counter
+		 * value, execute the handler.
+		 */
+		if (counter_sub(cc_data[chan].target_cc, counter + 1)
```
Comment (ankuns) [inline drivers/timer/nrf_rtc_timer.c:469]: Why is `+1` in `counter + 1` needed? can't following do the job?
```c
if (counter_sub(counter, cc_data[chan].target_cc) < COUNTER_HALF_SPAN)
```
This includes the case when counter == target_cc

File: include/drivers/timer/nrf_rtc_timer.h [hunk 5]
```diff
@@ -11,8 +11,12 @@
 extern "C" {
 #endif
 
+/** @brief Maximum allowed time span that is considered to be in the future.
+ */
+#define NRF_RTC_TIMER_MAX_SCHEDULE_SPAN (BIT(23) - 1)
+
 typedef void (*z_nrf_rtc_timer_compare_handler_t)(int32_t id,
-						uint32_t cc_value,
+						uint64_t cc_value,
```
Comment (ankuns) [inline include/drivers/timer/nrf_rtc_timer.h:19]: Let's decide if we are going to provide `cc_value` or `current_time` at which we real triggered.
We are going to provide `current_time` to not have to read timer again in callback.
Let's name this parameter `current_time`, `curr_time` or `now_time` but not `cc_value`.

**Review by ankuns — CHANGES_REQUESTED** [review]
First round

**Review by nordic-krch — COMMENTED** [review]
few comments.

**Review by e-rk — CHANGES_REQUESTED** [review]
First round of review.

**Review by ankuns — APPROVED** [review]

**Review by nordic-krch — COMMENTED** [review]
Couple of comments.

**Review by nordic-krch — APPROVED** [review]

**Review by nordic-krch — CHANGES_REQUESTED** [review]
Accidentally pressed approval.

**Review by jciupis — APPROVED** [review]

**Review by e-rk — APPROVED** [review]
Didn't review the latest changes, but I don't want github to block this PR any longer.

**Comment by nordic-krch** [comment]
Please update commit messages. There is a typo in `modules: hal_nordic: Align lptimer platform do nrf_rtc_timer ` and second commit mentions extending timeouts beyond half span.

**Review by nordic-krch — APPROVED** [review]
Looks good. One small thing: commit message `modules: hal_nordic: Align lptimer platform do nrf_rtc_timer` should probably be `modules: hal_nordic: Align lptimer platform to nrf_rtc_timer`

**Comment by carlescufi** [comment]
@pkwiek please take a look at the buildkite issues:

```
/workdir/zephyr/drivers/timer/nrf_rtc_timer.c:80:13: error: 'event_disable' defined but not used [-Werror=unused-function]
--
  | 80 \| static void event_disable(int32_t chan)
  | \|             ^~~~~~~~~~~~~
```

**Review by ankuns — APPROVED** [review]

File: include/drivers/timer/nrf_rtc_timer.h [hunk 6]
```diff
@@ -11,8 +11,25 @@
 extern "C" {
 #endif
 
+/** @brief Maximum allowed time span that is considered to be in the future.
+ */
+#define NRF_RTC_TIMER_MAX_SCHEDULE_SPAN (BIT(24) - 1)
```
Comment (nordic-krch) [inline include/drivers/timer/nrf_rtc_timer.h:16]: this is no longer true as mechanism for scheduling alarms longer than half span got removed.

Comment (jciupis) [reply include/drivers/timer/nrf_rtc_timer.h]: Fixed.

**Review by carlescufi — APPROVED** [review]

**Pull request merged**